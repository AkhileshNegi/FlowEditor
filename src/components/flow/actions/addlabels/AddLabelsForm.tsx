import { react as bindCallbacks } from 'auto-bind';
import Dialog, { ButtonSet } from 'components/dialog/Dialog';
import { ActionFormProps } from 'components/flow/props';
import AssetSelector from 'components/form/assetselector/AssetSelector';
import TypeList from 'components/nodeeditor/TypeList';
import { fakePropType } from 'config/ConfigProvider';
import * as React from 'react';
import { Asset } from 'store/flowContext';
import { AssetArrayEntry, FormState, mergeForm } from 'store/nodeEditor';
import { Required, validate } from 'store/validators';

import { initializeForm, stateToAction } from './helpers';

export interface AddLabelsFormState extends FormState {
  labels: AssetArrayEntry;
}

export const controlLabelSpecId = 'label';

export default class AddLabelsForm extends React.PureComponent<
  ActionFormProps,
  AddLabelsFormState
> {
  public static contextTypes = {
    assetService: fakePropType
  };

  constructor(props: ActionFormProps) {
    super(props);

    this.state = initializeForm(this.props.nodeSettings);
    bindCallbacks(this, {
      include: [/^on/, /^handle/]
    });
  }

  public handleSave(): void {
    const valid = this.handleLabelsChanged(this.state.labels.value!);

    if (valid) {
      const newAction = stateToAction(this.props.nodeSettings, this.state);
      this.props.updateAction(newAction);
      this.props.onClose(false);
    }
  }

  public handleLabelsChanged(selected: Asset[]): boolean {
    const updates: Partial<AddLabelsFormState> = {
      labels: validate('Labels', selected, [Required])
    };

    const updated = mergeForm(this.state, updates);
    this.setState(updated);
    return updated.valid;
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: 'Ok', onClick: this.handleSave },
      secondary: { name: 'Cancel', onClick: () => this.props.onClose(true) }
    };
  }

  public handleCreateAssetFromInput(input: string): any {
    return { name: input };
  }

  public handleLabelCreated(label: Asset): void {
    // update our store with our new group
    this.props.addAsset('labels', label);

    this.handleLabelsChanged(this.state.labels.value!.concat(label));
  }

  public render(): JSX.Element {
    const typeConfig = this.props.typeConfig;
    return (
      <Dialog title={typeConfig.name} headerClass={typeConfig.type} buttons={this.getButtons()}>
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        <p data-spec={controlLabelSpecId}>Select the labels to apply to the incoming message.</p>

        <AssetSelector
          name="Labels"
          placeholder="Enter the name of an existing label or create a new one"
          assets={this.props.assetStore.labels}
          entry={this.state.labels}
          searchable={true}
          multi={true}
          onChange={this.handleLabelsChanged}
          createPrefix="Create Label: "
          createAssetFromInput={this.handleCreateAssetFromInput}
          onAssetCreated={this.handleLabelCreated}
        />
      </Dialog>
    );
  }
}
