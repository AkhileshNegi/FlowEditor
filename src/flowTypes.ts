import {
    IsOptionUniqueHandler,
    IsValidNewOptionHandler,
    NewOptionCreatorHandler,
    PromptTextCreatorHandler
} from 'react-select';
import { Operators } from './config/operatorConfigs';
import AssetService from './services/AssetService';
import { Types } from './config/typeConfigs';

export interface Languages {
    [iso: string]: string;
}

export interface Endpoints {
    fields: string;
    groups: string;
    contacts: string;
    flows: string;
    activity: string;
    labels: string;
    simulateStart: string;
    simulateResume: string;
}

export interface FlowEditorConfig {
    assetService?: AssetService;
    localStorage: boolean;
    languages: { [iso: string]: string };
    endpoints: Endpoints;
    flow: string;
    path?: string;
}

export interface LocalizationMap {
    [lang: string]: {
        [uuid: string]: any;
    };
}

export interface FlowDefinition {
    localization: LocalizationMap;
    language: string;
    name: string;
    nodes: FlowNode[];
    uuid: string;
    _ui: UIMetaData;
}

export interface FlowNode {
    uuid: string;
    actions: Action[];
    exits: Exit[];
    router?: Router;
    wait?: Wait;
}

export interface Exit {
    uuid: string;
    name?: string;
    destination_node_uuid?: string;
}

export enum RouterTypes {
    switch = 'switch'
}

export interface Router {
    type: RouterTypes;
    result_name?: string;
}

export interface Case {
    uuid: string;
    type: Operators;
    exit_uuid: string;
    arguments?: string[];
}

export interface SwitchRouter extends Router {
    cases: Case[];
    operand: string;
    default_exit_uuid: string;
}

export enum WaitTypes {
    exp = 'exp',
    group = 'group',
    msg = 'msg',
    flow = 'flow'
}

export interface Wait {
    type: WaitTypes;
    timeout?: number;
}

export interface Group {
    uuid: string;
    name: string;
}

export enum Methods {
    GET = 'GET',
    POST = 'POST',
    PUT = 'PUT'
}

export interface ChangeGroups extends Action {
    groups: Group[];
}

export interface Field {
    key: string;
    name: string;
}

export interface Label {
    uuid: string;
    name: string;
}

export interface Action {
    type: Types;
    uuid: string;
}

export interface SetContactField extends Action {
    field: Field;
    value: string;
}

export interface SetContactName extends Action {
    type: Types.set_contact_name;
    name: string;
}

export type SetContactProperty = SetContactName;

export type SetContactAttribute = SetContactField | SetContactProperty;

// tslint:disable-next-line:no-empty-interface
export interface Missing extends Action {}

export interface SendMsg extends Action {
    text: string;
    all_urns?: boolean;
}

export interface AddLabels extends Action {
    labels: Label[];
}

export interface SetPreferredChannel extends Action {
    language: string;
}

export interface SendEmail extends Action {
    subject: string;
    body: string;
    addresses: string[];
}

export interface SetRunResult extends Action {
    name: string;
    value: string;
    category?: string;
}

export interface Headers {
    [name: string]: string;
}

export interface CallWebhook extends Action {
    url: string;
    method: Methods;
    body?: string;
    headers?: Headers;
}

export interface StartFlow extends Action {
    flow: {
        name: string;
        uuid: string;
    };
}

export interface UIMetaData {
    nodes: { [key: string]: UINode };
    languages: Array<{ [iso: string]: string }>;
}

export interface FlowPosition {
    left: number;
    top: number;
    right?: number;
    bottom?: number;
}

export interface Dimensions {
    width: number;
    height: number;
}

export enum UINodeTypes {
    split = 'split',
    subflow = 'subflow',
    webhook = 'webhook'
}

export interface UINode {
    position: FlowPosition;
    // ui type, used for split by expression, contact field, etc
    type?: UINodeTypes | Types;
}

export interface StickyNote {
    position: FlowPosition;
    title: string;
    body: string;
    color?: string;
}

export interface UIMetaData {
    nodes: { [key: string]: UINode };
    stickies: { [key: string]: StickyNote };
}

export type AnyAction =
    | Action
    | ChangeGroups
    | SetContactField
    | SetContactProperty
    | SetRunResult
    | SendMsg
    | SetPreferredChannel
    | SendEmail
    | CallWebhook
    | StartFlow;

export enum ContactProperties {
    UUID = 'UUID',
    'Created By' = 'Created By',
    'Modified By' = 'Modified By',
    Org = 'Org',
    Name = 'Name',
    Language = 'Language',
    Timezone = 'Timezone',
    Email = 'Email',
    Mailto = 'Mailto',
    Phone = 'Phone',
    Groups = 'Groups',
    Facebook = 'Facebook',
    Telegram = 'Telegram'
}

export enum ResultType {
    flow = 'flow',
    field = 'field',
    group = 'group',
    label = 'label'
}

export enum ValueType {
    text = 'text',
    numeric = 'numeric',
    datetime = 'datetime',
    state = 'state',
    district = 'district',
    ward = 'ward'
}

export interface CreateOptions {
    promptTextCreator?: PromptTextCreatorHandler;
    newOptionCreator?: NewOptionCreatorHandler;
    isValidNewOption?: IsValidNewOptionHandler;
    isOptionUnique?: IsOptionUniqueHandler;
    createNewOption?: NewOptionCreatorHandler;
    createPrompt?: string;
}

export enum StartFlowArgs {
    Complete = 'C',
    Expired = 'E'
}

export enum StartFlowExitNames {
    Complete = 'Complete',
    Expired = 'Expired'
}

export enum WebhookExitNames {
    Success = 'Success',
    Failure = 'Failed',
    Unreachable = 'Unreachable'
}
