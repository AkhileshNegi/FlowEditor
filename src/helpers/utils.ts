import { ComponentClass, SFC, ReactElement } from 'react';
import { ShallowWrapper, ReactWrapper } from 'enzyme';
import { FlowDefinition } from '../flowTypes';

const request = require('sync-request');

const XRegExp = require('xregexp');

const SNAKED_CHARS = XRegExp('[^\\p{Letter}\\d]+');

interface BoolMap {
    [key: string]: boolean;
}

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {object} A map of each string
 */
export function toBoolMap(array: string[]): BoolMap {
    return array.reduce((map: BoolMap, item: string) => {
        map[item] = true;
        return map;
    }, {});
}

/**
 * Inserts commas into numbers where appropriate for better readability
 * @param {number} value - A number, e.g. 10000
 * @returns {string} A comma-separated string, e.g. 10,000
 */
export function addCommas(value: number): string {
    return value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export function snakify(value: string): string {
    return XRegExp.replace(value.toLowerCase().trim(), SNAKED_CHARS, '_', 'all');
}

/**
 * Returns a React component's name so we can attach it to a HOC's displayName property and view it in RDT
 * @param {ComponentClass | SFC} Component - A React component
 * @returns {string} The component's name
 */
export function getDisplayName(HOCName: string, Component: ComponentClass | SFC): string {
    const ComponentDisplayName = Component.displayName || Component.name || 'Component';
    return `${HOCName}(${ComponentDisplayName})`;
}

function getFlow(name: string): FlowDefinition {
    const definition = request('GET', 'base/test_flows/' + name + '.json').getBody();
    return JSON.parse(definition) as FlowDefinition;
}

export function getFavorites(): FlowDefinition {
    return getFlow('favorites');
}

export function getTest(): FlowDefinition {
    return getFlow('a4f64f1b-85bc-477e-b706-de313a022979');
}

export function dump(object: any) {
    console.log(JSON.stringify(object, null, 1));
}

export const DATA_SPEC_ATTRIBUTE_NAME: string = 'data-spec';

/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} matching DOM components
 */
export function getSpecWrapper(
    componentWrapper: ReactWrapper | ShallowWrapper,
    specName: string
): ShallowWrapper | ReactWrapper {
    return componentWrapper.find(`[${DATA_SPEC_ATTRIBUTE_NAME}="${specName}"]`);
}

/**
 * Returns true if a given UUID matches v4 format
 * @param {string} uuid - A version 4 UUID (no braces, uppercase OK)
 * @returns {boolean}
 */
export function validUUID(uuid: string): boolean {
    const regex = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;
    return regex.test(uuid);
}
