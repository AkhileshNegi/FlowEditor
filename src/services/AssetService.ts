// tslint:disable:max-classes-per-file
import { v4 as generateUUID } from 'uuid';
import { Endpoints, FlowEditorConfig, ContactProperties, Group } from '../flowTypes';
import axios, { AxiosResponse } from 'axios';
import { FlowComponents } from '../store/helpers';
import { dump } from '../utils';

export enum AssetType {
    Channel = 'channel',
    Flow = 'flow',
    Group = 'group',
    Property = 'property',
    Field = 'field'
}

export interface Asset {
    id: string;
    name: string;
    type: AssetType;

    isNew?: boolean;
    content?: any;
}

enum IdProperty {
    UUID = 'uuid',
    Key = 'key',
    ID = 'id'
}

enum NameProperty {
    Name = 'name'
}

enum SimAssetType {
    Flow = 'flow',
    Fields = 'field_set',
    Groups = 'group_set',
    Channels = 'channel_set'
}

interface SimAsset {
    type: SimAssetType;
    url: string;
    content: any;
}

export const getBaseURL = (): string => {
    const location = window.location;
    return (
        location.protocol + '//' + location.hostname + (location.port ? ':' + location.port : '')
    );
};

export class Assets {
    public endpoint: string;
    private localStorage: boolean;
    protected idProperty: IdProperty;
    protected assetType: AssetType;
    protected assets: { [id: string]: Asset } = {};

    constructor(endpoint: string, localStorage: boolean) {
        this.localStorage = localStorage;
        this.endpoint = endpoint;
        this.idProperty = IdProperty.UUID;
    }

    public get(id: string): Promise<Asset> {
        const existing = this.assets[id];
        if (existing) {
            return new Promise<Asset>(resolve => {
                resolve(existing);
            });
        }

        return new Promise<Asset>((resolve, reject) => {
            const url = `${this.endpoint}?uuid=${id}`;
            axios
                .get(url)
                .then((response: AxiosResponse) => {
                    const ob = response.data.results[0];
                    const asset = {
                        id: ob.uuid,
                        name: ob.name,
                        type: this.assetType,
                        content: ob.definition
                    };
                    if (this.localStorage) {
                        this.assets[id] = asset;
                    }
                    return resolve(asset);
                })
                .catch(error => reject(error));
        });
    }

    private searchLocalItems(term: string): Asset[] {
        // search our local items first
        const matches: Asset[] = [];
        Object.keys(this.assets).map((key: string) => {
            const asset = this.assets[key];
            if (this.matches(term, asset.name)) {
                matches.push(asset);
            }
        });
        return matches;
    }

    public search(term: string): Promise<Asset[]> {
        const matches = this.searchLocalItems(term);

        // then query against our endpoint to add to that list
        let url = this.endpoint;
        if (term) {
            url += '?query=' + encodeURIComponent(term);
        }

        return axios.get(url).then((response: AxiosResponse) => {
            for (const result of response.data.results) {
                if (this.matches(term, result.name)) {
                    matches.push({
                        name: result.name,
                        id: result[this.idProperty],
                        type: this.assetType
                    });
                }
            }
            return matches;
        });
    }

    public matches(query: string, check: string): boolean {
        if (!check) {
            return false;
        }

        if (query.length === 0) {
            return true;
        }

        return (
            check
                .toLocaleLowerCase()
                .trim()
                .indexOf(query.toLocaleLowerCase().trim()) > -1
        );
    }

    public add(result: Asset): void {
        if (this.localStorage) {
            this.assets[result.id] = result;
        }
    }

    public addAll(results: Asset[]): void {
        results.map((result: Asset) => {
            this.add(result);
        });
    }

    public update(uuid: string, content: any): Promise<Asset> {
        const result = new Promise<Asset>((resolve, reject) => {
            if (this.localStorage) {
                const asset = this.assets[uuid];
                asset.content = content;
                resolve(asset);
            } else {
                const url = `${this.endpoint}?uuid=${uuid}`;
                axios
                    .post(url, content)
                    .then((response: AxiosResponse) => {
                        const ob = response.data.results[0];
                        const asset = {
                            id: ob.uuid,
                            name: ob.name,
                            type: this.assetType,
                            content: ob.definition
                        };
                        if (this.localStorage) {
                            this.assets[uuid] = asset;
                        }
                        return resolve(asset);
                    })
                    .catch(error => reject(error));
            }
        });

        return result;
    }

    public getAssetSet(): any[] {
        const assets: any[] = [];
        Object.keys(this.assets).map((key: string) => {
            const asset = this.assets[key];
            if (asset.type !== AssetType.Property) {
                assets.push({
                    [this.idProperty]: asset.id,
                    name: asset.name,
                    ...asset.content
                });
            }
        });
        return assets;
    }

    public getAssetContents(type: SimAssetType): SimAsset[] {
        const assets: SimAsset[] = [];
        Object.keys(this.assets).map((key: string) => {
            const asset = this.assets[key];
            assets.push({
                type,
                url: getBaseURL() + this.endpoint + '/' + asset.id + '/',
                content: asset.content
            });
        });

        return assets;
    }
}

class GroupAssets extends Assets {
    constructor(endpoint: string, localStorage: boolean) {
        super(endpoint, localStorage);
        this.idProperty = IdProperty.UUID;
        this.assetType = AssetType.Group;
    }
}

class FieldAssets extends Assets {
    public static CONTACT_PROPERTIES: Asset[] = [
        {
            name: ContactProperties.Name,
            id: ContactProperties.Name.toLowerCase(),
            type: AssetType.Property
        }
        /*{ 
            name: ContactProperties.Language, 
            id: ContactProperties.Language.toLowerCase(), 
            type: AttributeType.property 
        }*/
    ];

    constructor(endpoint: string, localStorage: boolean) {
        super(endpoint, localStorage);
        this.idProperty = IdProperty.Key;
        this.assetType = AssetType.Field;

        FieldAssets.CONTACT_PROPERTIES.map((result: Asset) => {
            this.assets[result.id] = result;
        });
    }
}

class ChannelAssets extends Assets {
    constructor(endpoint: string, localStorage: boolean) {
        super(endpoint, localStorage);
        this.idProperty = IdProperty.UUID;
        this.assetType = AssetType.Channel;

        const uuid = '6934255e-446b-40d3-b2ca-c5f801bd2278';
        this.assets[uuid] = {
            id: uuid,
            name: 'Simulator',
            type: AssetType.Channel,
            content: {
                address: '+12065550000',
                schemes: ['tel'],
                roles: ['send', 'receive']
            }
        };
    }
}

class FlowAssets extends Assets {
    constructor(endpoint: string, localStorage: boolean) {
        super(endpoint, localStorage);
        this.idProperty = IdProperty.UUID;
        this.assetType = AssetType.Flow;
    }
}

export default class AssetService {
    private channels: Assets;
    private groups: Assets;
    private fields: Assets;
    private flows: FlowAssets;

    constructor(config: FlowEditorConfig) {
        this.groups = new GroupAssets(config.endpoints.groups, config.localStorage);
        this.fields = new FieldAssets(config.endpoints.fields, config.localStorage);
        this.flows = new FlowAssets(config.endpoints.flows, config.localStorage);

        // channels are always mocked for local
        this.channels = new ChannelAssets('/channels', true);
    }

    public addFlowComponents(flowComponents: FlowComponents): void {
        this.groups.addAll(flowComponents.groups);
        this.fields.addAll(flowComponents.fields);
    }

    public getFlowAssets(): Assets {
        return this.flows;
    }

    public getGroupAssets(): Assets {
        return this.groups;
    }

    public getFieldAssets(): Assets {
        return this.fields;
    }

    public getSimulationAssets(): any {
        let simAssets: SimAsset[] = [];
        const base = getBaseURL();
        // our group set asset
        simAssets.push({
            type: SimAssetType.Groups,
            url: base + this.groups.endpoint + '/',
            content: this.groups.getAssetSet()
        });

        // our fields
        simAssets.push({
            type: SimAssetType.Fields,
            url: base + this.fields.endpoint + '/',
            content: this.fields.getAssetSet()
        });

        // our channels
        simAssets.push({
            type: SimAssetType.Channels,
            url: base + this.channels.endpoint + '/',
            content: this.channels.getAssetSet()
        });

        // our flows
        simAssets = simAssets.concat(this.flows.getAssetContents(SimAssetType.Flow));

        const payload = {
            assets: simAssets,
            asset_server: {
                type_urls: {
                    flow: base + this.flows.endpoint + '/{uuid}/',
                    field_set: base + this.fields.endpoint + '/',
                    channel_set: base + this.channels.endpoint + '/',
                    group_set: base + this.groups.endpoint + '/'
                }
            }
        };

        return payload;
    }
}
