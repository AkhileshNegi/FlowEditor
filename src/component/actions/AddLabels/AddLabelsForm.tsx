import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';

import { ConfigProviderContext } from '../../../config';
import { fakePropType } from '../../../config/ConfigProvider';
import { Types } from '../../../config/typeConfigs';
import { AddLabels, Label } from '../../../flowTypes';
import AssetService, { Asset, AssetType } from '../../../services/AssetService';
import LabelsElement from '../../form/LabelsElement';
import { mapAssetsToGroups, mapGroupsToAssets } from '../ChangeGroups/helpers';

export interface AddLabelsFormProps {
    action: AddLabels;
    onBindWidget: (ref: any) => void;
    updateAction: (action: AddLabels) => void;
}

export const mapLabelsToAssets = (labels: Label[]): Asset[] =>
    labels.map(({ name, uuid }) => ({ name, id: uuid, type: AssetType.Label }));

export const mapAssetsToLabels = (searchResults: Asset[]): Label[] =>
    searchResults.map(({ id, name }) => ({ uuid: id, name }));

export const createNewAddLabelAction = ({ uuid }: AddLabels, labels: Asset[]) => ({
    uuid,
    type: Types.add_input_labels,
    labels: mapAssetsToLabels(labels)
});

export const LABEL = ' Select the label(s) to apply to the incoming message.';
export const PLACEHOLDER = 'Enter the name of an existing label or create a new one';

export const controlLabelSpecId = 'label';

export default class AddLabelsForm extends React.PureComponent<AddLabelsFormProps> {
    public static contextTypes = {
        assetService: fakePropType
    };

    constructor(props: AddLabelsFormProps, context: ConfigProviderContext) {
        super(props);

        bindCallbacks(this, {
            include: [/^on/]
        });
    }

    public onValid(widgets: { [name: string]: any }): void {
        const { state: { labels } } = widgets.Labels;
        const newAction = createNewAddLabelAction(this.props.action, labels);

        this.props.updateAction(newAction);
    }

    private getLabels(): Asset[] {
        if (!this.props.action.labels.length) {
            return [];
        }
        return mapLabelsToAssets(this.props.action.labels);
    }

    public render(): JSX.Element {
        return (
            <>
                <p data-spec={controlLabelSpecId}>{LABEL}</p>
                <LabelsElement
                    ref={this.props.onBindWidget}
                    name="Labels"
                    placeholder={PLACEHOLDER}
                    assets={this.context.assetService.getLabelAssets()}
                    labels={this.getLabels()}
                />
            </>
        );
    }
}
