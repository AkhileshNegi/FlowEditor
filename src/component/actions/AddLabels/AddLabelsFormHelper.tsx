import { FormHelper, Types } from '../../../config/typeConfigs';
import { AddLabels } from '../../../flowTypes';
import { Asset, AssetType } from '../../../services/AssetService';
import { AddLabelsFormState } from '../../../store/nodeEditor';

export class AddLabelsFormHelper implements FormHelper {
    public actionToState(action: AddLabels): AddLabelsFormState {
        if (action) {
            return {
                type: action.type,
                labels: {
                    value: action.labels.map(label => {
                        return { id: label.uuid, name: label.name, type: AssetType.Label };
                    })
                },
                valid: true
            };
        }

        return {
            type: Types.add_input_labels,
            labels: { value: [] },
            valid: false
        };
    }

    public stateToAction(actionUUID: string, formState: AddLabelsFormState): AddLabels {
        return {
            labels: this.getAsset(formState.labels.value, AssetType.Label),
            type: formState.type,
            uuid: actionUUID
        };
    }

    private getAsset(assets: Asset[], type: AssetType): any[] {
        return assets.filter((asset: Asset) => asset.type === type).map((asset: Asset) => {
            return { uuid: asset.id, name: asset.name };
        });
    }
}
