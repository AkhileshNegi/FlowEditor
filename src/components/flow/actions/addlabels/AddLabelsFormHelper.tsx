import { FormHelper, Types } from '~/config/typeConfigs';
import { AddLabels } from '~/flowTypes';
import { Asset, AssetType } from '~/services/AssetService';
import { AddLabelsFormState, NodeEditorSettings } from '~/store/nodeEditor';

export class AddLabelsFormHelper implements FormHelper {
    public initializeForm(settings: NodeEditorSettings): AddLabelsFormState {
        if (settings.originalAction && settings.originalAction.type === Types.add_input_labels) {
            const action = settings.originalAction as AddLabels;
            return {
                labels: {
                    value: action.labels.map(label => {
                        return { id: label.uuid, name: label.name, type: AssetType.Label };
                    })
                },
                valid: true
            };
        }

        return {
            labels: { value: [] },
            valid: false
        };
    }

    public stateToAction(actionUUID: string, formState: AddLabelsFormState): AddLabels {
        return {
            type: Types.add_input_labels,
            labels: this.getAsset(formState.labels.value, AssetType.Label),
            uuid: actionUUID
        };
    }

    private getAsset(assets: Asset[], type: AssetType): any[] {
        return assets.filter((asset: Asset) => asset.type === type).map((asset: Asset) => {
            return { uuid: asset.id, name: asset.name };
        });
    }
}
