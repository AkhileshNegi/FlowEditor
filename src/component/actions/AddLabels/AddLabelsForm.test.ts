import { Label } from '~/flowTypes';
import { AssetType } from '~/services/AssetService';
import { composeComponentTestUtils } from '~/testUtils';
import { createAddLabelsAction } from '~/testUtils/assetCreators';
import { setTrue } from '~/utils';

import { AddLabelsForm, AddLabelsFormProps } from './AddLabelsForm';
import { AddLabelsFormHelper } from './AddLabelsFormHelper';

const { assets: labels } = require('~/test/assets/labels.json') as {
    assets: Label[];
};

const formHelper = new AddLabelsFormHelper();
const action = createAddLabelsAction(labels);
const baseProps: AddLabelsFormProps = {
    action,
    updateAction: jest.fn(),
    updateAddLabelsForm: jest.fn(),
    form: formHelper.initializeForm({ originalNode: null, originalAction: action }),
    formHelper
};

const { setup, spyOn } = composeComponentTestUtils(AddLabelsForm, baseProps);

describe(AddLabelsForm.name, () => {
    describe('render', () => {
        it('should pass LabelsElement labels if they exist on the action', () => {
            const { wrapper } = setup();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('onValid', () => {
        it('should update action', () => {
            const emptyAction = createAddLabelsAction([]);
            const { wrapper, instance, props } = setup(true, {
                action: { $set: emptyAction },
                form: {
                    $set: formHelper.initializeForm({
                        originalNode: null,
                        originalAction: emptyAction
                    })
                },
                updateAction: { $set: jest.fn() },
                updateAddLabelsForm: { $set: jest.fn().mockReturnValue(setTrue) }
            });

            instance.handleLabelChange([
                { id: 'label0', name: 'Updated Label', type: AssetType.Label }
            ]);

            instance.onValid();
            expect(props.updateAction).toHaveBeenCalledTimes(1);
            expect(props.updateAction).toHaveBeenCalledWith({
                labels: [],
                type: 'add_input_labels',
                uuid: 'labels-action-uuid-0'
            });
        });
    });
});
