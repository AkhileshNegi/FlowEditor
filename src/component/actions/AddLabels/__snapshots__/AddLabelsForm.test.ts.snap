// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddLabelsForm render should pass LabelsElement an empty labels array if they do not exist on the action 1`] = `
<React.Fragment>
  <p
    data-spec="label"
  >
     Select the label(s) to apply to the incoming message.
  </p>
  <LabelsElement
    assets={
      LabelAssets {
        "assetType": "label",
        "assets": Object {},
        "endpoint": "/assets/labels.json",
        "idProperty": "uuid",
        "localStorage": true,
      }
    }
    labels={Array []}
    name="Labels"
    placeholder="Enter the name of an existing label or create a new one"
    required={true}
    searchPromptText="Invalid label"
  />
</React.Fragment>
`;

exports[`AddLabelsForm render should pass LabelsElement labels if they exist on the action 1`] = `
<React.Fragment>
  <p
    data-spec="label"
  >
     Select the label(s) to apply to the incoming message.
  </p>
  <LabelsElement
    assets={
      LabelAssets {
        "assetType": "label",
        "assets": Object {},
        "endpoint": "/assets/labels.json",
        "idProperty": "uuid",
        "localStorage": true,
      }
    }
    labels={
      Array [
        Object {
          "id": "232b99d1-579f-4d7f-b750-4b034927e9f8",
          "name": "Help",
          "type": "label",
        },
        Object {
          "id": "ab58dffa-98da-40ba-bebc-489ec6ff7170",
          "name": "Feedback",
          "type": "label",
        },
      ]
    }
    name="Labels"
    placeholder="Enter the name of an existing label or create a new one"
    required={true}
    searchPromptText="Invalid label"
  />
</React.Fragment>
`;
